File: .claude/skills/event-aggregator/src/commands/rotate.ts
import * as db from "../db.ts";
import { norm } from "../rank.ts";

export const MIN_RUN = 3;

export type RotateOpts = { slugs: string[]; minRun: number; all: boolean };

type Block = { title: string; venue: string; dates: string[]; firstSeen: string };

function blocks(conn: ReturnType<typeof db.connect>, slug: string, gone: boolean) {
  const out = new Map<string, Block>();
  for (const r of db.all(
    conn,
    "SELECT title, venue, start_at, first_seen FROM events WHERE source=? AND gone=?",
    [slug, gone ? 1 : 0],
  )) {
    const title = String(r["title"] ?? "");
    const venue = String(r["venue"] ?? "");
    const k = `${norm(title)} ${venue}`;
    const b = out.get(k) ?? { title, venue, dates: [], firstSeen: "9999" };
    b.dates.push(String(r["start_at"]).slice(0, 10));
    const fs = String(r["first_seen"]);
    if (fs < b.firstSeen) b.firstSeen = fs;
    out.set(k, b);
  }
  return out;
}

/** Diff against the run BEFORE the latest one, so a fresh scrape does not read
 *  as "everything is new". */
function previousRun(conn: ReturnType<typeof db.connect>, slug: string): string | null {
  const r = db.one(
    conn,
    "SELECT started_at FROM runs WHERE slug=? AND ok=1 ORDER BY id DESC LIMIT 1 OFFSET 1",
    [slug],
  );
  return r ? String(r["started_at"]) : null;
}

export function main(o: RotateOpts): number {
  const conn = db.connect();
  const bl = db.blacklist();

  const todo = o.slugs.length
    ? o.slugs
    : db.all(conn, "SELECT slug FROM scrapers WHERE status='built' ORDER BY slug")
      .map((r) => String(r["slug"]))
      .filter((s) => !bl.sources.has(s));

  let shown = 0;
  for (const slug of todo) {
    const prev = previousRun(conn, slug);
    const live = blocks(conn, slug, false);
    const dead = blocks(conn, slug, true);

    const entered = [...live.values()]
      .filter((b) => b.dates.length >= o.minRun && prev && b.firstSeen > prev);
    const left = [...dead.entries()]
      .filter(([k, b]) => b.dates.length >= o.minRun && !live.has(k))
      .map(([, b]) => b);

    if (!entered.length && !left.length && !o.all) continue;
    shown++;

    const note = prev ? "" : "  (only one run on record - nothing to diff yet)";
    console.log(`== ${slug} ==${note}`);
    for (const b of entered) {
      const from = b.dates.slice().sort()[0] ?? "?";
      console.log(
        `   + ${b.title.slice(0, 46).padEnd(46)} ` +
        `${String(b.dates.length).padStart(2)} dates from ${from}`,
      );
    }
    for (const b of left) {
      console.log(
        `   - ${b.title.slice(0, 46).padEnd(46)} ` +
        `${String(b.dates.length).padStart(2)} dates dropped`,
      );
    }
    if (!entered.length && !left.length) console.log("   (no change)");
    console.log("");
  }

  if (!shown) console.log(`no rotation changes (run length >= ${o.minRun})`);
  return 0;
}

File: .claude/skills/event-aggregator/rotate.py
import argparse
import re
import sys
from collections import defaultdict

import db

MIN_RUN = 3


def norm(t):
    t = re.sub(r"\(.*?\)|\b-?2D-?\b|OmeU|eOmU|OmU", "", t or "")
    return re.sub(r"[^a-z0-9]", "", t.lower())[:40]


def blocks(conn, slug, gone):
    out = defaultdict(list)
    for r in conn.execute(
            "SELECT title, venue, start_at, first_seen FROM events"
            " WHERE source=? AND gone=?", (slug, 1 if gone else 0)):
        out[(norm(r["title"]), r["venue"] or "")].append(r)
    return out


def since_stamp(conn, slug):
    row = conn.execute(
        "SELECT started_at FROM runs WHERE slug=? AND ok=1"
        " ORDER BY id DESC LIMIT 1 OFFSET 1", (slug,)).fetchone()
    return row["started_at"] if row else None


def report(conn, slug, min_run):
    prev = since_stamp(conn, slug)
    live = blocks(conn, slug, gone=False)
    dead = blocks(conn, slug, gone=True)

    entered = []
    for k, v in live.items():
        if len(v) < min_run:
            continue
        first = min(r["first_seen"] for r in v)
        if prev and first > prev:
            entered.append((v[0]["title"], v[0]["venue"], len(v),
                            min(r["start_at"][:10] for r in v)))
    left = [(v[0]["title"], v[0]["venue"], len(v))
            for k, v in dead.items() if len(v) >= min_run and k not in live]
    return prev, entered, left


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("slugs", nargs="*")
    ap.add_argument("--min-run", type=int, default=MIN_RUN)
    ap.add_argument("--all", action="store_true",
                    help="include sources with no rotation change")
    a = ap.parse_args()

    conn = db.connect()
    srcs, _, _ = db.blacklist()
    todo = a.slugs or [r["slug"] for r in conn.execute(
        "SELECT slug FROM scrapers WHERE status='built' ORDER BY slug")
        if r["slug"] not in srcs]

    shown = 0
    for slug in todo:
        prev, entered, left = report(conn, slug, a.min_run)
        if not entered and not left and not a.all:
            continue
        shown += 1
        note = "" if prev else "  (only one run on record - nothing to diff yet)"
        print("== %s ==%s" % (slug, note))
        for t, v, n, d in entered:
            print("   + %-46s %2d dates from %s" % (t[:46], n, d))
        for t, v, n in left:
            print("   - %-46s %2d dates dropped" % (t[:46], n))
        if not entered and not left:
            print("   (no change)")
        print()

    if not shown:
        print("no rotation changes (run length >= %d)" % a.min_run)
    return 0


if __name__ == "__main__":
    sys.exit(main())
